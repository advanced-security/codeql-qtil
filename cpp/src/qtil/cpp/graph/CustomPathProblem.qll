private import qtil.locations.CustomPathProblem
private import qtil.cpp.locations.Locatable
private import cpp
// Import the C++ specific configuration for making custom path problems.
import PathProblem<Location, CppLocatableConfig>
