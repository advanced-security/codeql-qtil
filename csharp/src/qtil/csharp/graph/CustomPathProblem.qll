private import qtil.locations.CustomPathProblem
private import qtil.csharp.locations.Locatable
private import csharp
// Import the C# specific configuration for making custom path problems.
import PathProblem<Location, CsharpLocatableConfig>
