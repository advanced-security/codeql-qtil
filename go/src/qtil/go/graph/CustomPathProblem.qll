private import qtil.locations.CustomPathProblem
private import qtil.go.locations.Locatable
private import go
// Import the Go specific configuration for making custom path problems.
import PathProblem<DbLocation, GoLocatableConfig>
