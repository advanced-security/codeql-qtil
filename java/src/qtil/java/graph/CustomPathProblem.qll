private import qtil.locations.CustomPathProblem
private import qtil.java.locations.Locatable
private import java
// Import the Java specific configuration for making custom path problems.
import PathProblem<Location, JavaLocatableConfig>
