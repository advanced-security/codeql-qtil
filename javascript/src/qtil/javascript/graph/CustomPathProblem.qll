private import qtil.locations.CustomPathProblem
private import qtil.javascript.locations.Locatable
private import javascript
// Import the javascript specific configuration for making custom path problems.
import PathProblem<DbLocation, JavascriptLocatableConfig>
