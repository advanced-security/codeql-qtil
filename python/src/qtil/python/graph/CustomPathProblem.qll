private import qtil.locations.CustomPathProblem
private import qtil.python.locations.Locatable
private import python
// Import the python specific configuration for making custom path problems.
import PathProblem<Location, PythonLocatableConfig>
