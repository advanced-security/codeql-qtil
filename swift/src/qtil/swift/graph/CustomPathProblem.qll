private import qtil.locations.CustomPathProblem
private import qtil.swift.locations.Locatable
private import swift
// Import the swift specific configuration for making custom path problems.
import PathProblem<Location, SwiftLocatableConfig>
