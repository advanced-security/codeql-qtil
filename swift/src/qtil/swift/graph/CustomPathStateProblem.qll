private import qtil.locations.CustomPathStateProblem
private import qtil.swift.locations.Locatable
private import swift
// Import the swift specific configuration for making custom path state problems.
import PathStateProblem<Location, SwiftLocatableConfig>
