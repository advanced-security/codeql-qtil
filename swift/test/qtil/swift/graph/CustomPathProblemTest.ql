// Minimal test to make sure CustomPathProblem works with swift code.
import swift
import qtil.swift.graph.CustomPathProblem

module CallGraphPathProblemConfig implements CustomPathProblemConfigSig {
  class Node = VarDecl;

  predicate start(Node n) { n.getName() = "start" }

  predicate end(Node n) {
    n.getName() = "end"
  }

  predicate edge(VarDecl a, VarDecl b) {
    exists(AssignExpr assign |
      assign.getDest() = a.getAnAccess() and
      assign.getSource() = b.getAnAccess()
    )
  }
}

import CustomPathProblem<CallGraphPathProblemConfig>

from VarDecl start, VarDecl end
where problem(start, end)
select start, end, "Path from $@ to $@.", start.getName(), start, end.getName(), end
