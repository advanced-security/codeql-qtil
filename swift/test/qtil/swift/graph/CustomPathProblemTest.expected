edges
| test.swift:2:9:2:9 | mid | test.swift:4:9:4:9 | end |  |  |
| test.swift:3:9:3:9 | start | test.swift:2:9:2:9 | mid |  |  |
nodes
| test.swift:2:9:2:9 | mid | semmle.label | mid |
| test.swift:3:9:3:9 | start | semmle.label | start |
| test.swift:4:9:4:9 | end | semmle.label | end |
#select
| test.swift:3:9:3:9 | start | test.swift:4:9:4:9 | end | Path from $@ to $@. | start | test.swift:3:9:3:9 | start | end | test.swift:4:9:4:9 | end |
